apps/email/js/cards/cmp/autocomplete.js

```diff
@@ -0,0 +1,565 @@
+      document.addEventListener('click', this.onDocumentEvent, true);
+      document.addEventListener('focus', this.onDocumentEvent, true);
+    },
+
+    /**
+     * Custom element callback when element is removed from the document.
+     */
+    detachedCallback: function() {
+      window.removeEventListener('resize', this.onResize, false);
+      document.removeEventListener('click', this.onDocumentClick, true);
+      document.addEventListener('focus', this.onDocumentEvent, true);
+    },
+
+    /**
+     * Triggered by input focus events.
+     */
+    onFocus: function(event) {
+      this.inputNode = event.target;
+
+      // If existing label node, be sure to remove the active class from it. In
+      // cases where the focus goes from one input to the other, hide() does not
+      // run to clear this state.
+      if (this.inputLabelNode) {
+        this.inputLabelNode.classList.remove('cmp-autocomplete-active');
+      }
+
+      this.inputLabelNode = findAncestorWithClass(this.inputNode,
+                                                'cmp-autocomplete-input-label');
+      this.inputLabelNode.classList.add('cmp-autocomplete-active');
+
+      this.query = this.inputNode.value.trim();
+      this.runThrottledQuery();
+    },
+
+    /**
+     * Triggered by any click or focus in the document. Used to determine if the
+     * autocomplete should be hidden.
+     */
+    onDocumentEvent: function(event) {
+      var target = event.target;
+
+      // If no classList, then it is not an element, and should be discarded.
+      if (!target.classList) {
+        this.hide();
+        return;
+      }
+
+      // If an autocomplete input, do not do anything, handled by other code in
+      // this module.
+      if (target.classList.contains('cmp-autocomplete-input')) {
+        return;
+      }
+
+      // Also ignore if it is part of an autocomplete-item.
+      if (getAutocompleteItem(target)) {
+        return;
+      }
+      this.hide();
+    },
+
+    onKeyDown: function(event) {
+      // If the down arrow is pressed, then focus on the first autocomplete
+      // item, if there is one.
+      if (event.key === 'ArrowDown' && !this.isHidden) {
+        var node = this.list.querySelector('.cmp-autocomplete-item-link');
+        if (node) {
+          node.focus();
+          event.stopPropagation();
+          event.preventDefault();
+        }
+      }
+    },
+
+    /**
+     * Triggered by input element input events.
+     */
+    onInput: function(event) {
+      this.query = this.inputNode.value.trim();
+      this.runThrottledQuery();
+    },
+
+    /**
+     * Handles window resizes. Could be triggered by keyboard size changes.
+     */
+    onResize: function(event) {
+      if (this.inputNode && !this.isHidden) {
+        this.positionElements();
+      }
+    },
+
+    onMatchKeyDown: function(event) {
+      if (event.key === 'Enter' && !this.isHidden) {
+        this.onMatchClick(event);
+      }
```

Comment:
I feel like this needs ArrowDown and ArrowUp logic of its own or disclaimers.  It's super weird to be able to hit arrow down to focus a match and then seem like you're trapped there.

I realize you did mention that tab works in your bullet change list, but it seems odd to me to support down once and then have it stop working.  This is verging on feature creep quite a bit, of course.  I'd be fine if you have the onKeyDown support explicitly disclaim keyboard navigation as a feature.  Maybe we should just mess with the tabstop ordering so only tab ever works?  (On b2g-desktop I end up tabbing myself into the bcc box from the cc box.)

So I did try tab based on your guidance, but that ended up extra weird.  When my tab focus left the autocomplete items, they didn't go away.  I guess I ended up in the bcc line and then the subject and then the body.  And then, somewhat hilariously, I guess I managed to tab into the "search upsell" box on the message list screen, because all of a sudden the message list search card transitioned onto the screen.  (I'm guessing whatever selector magic we're trying to use to stop text input from happening on other cards is not enough to negate its tab order?)